File: C++/I/11-ZOO/6/dragdroplabel.cpp
#include "dragdroplabel.h"
#include <QDrag>
#include <QMimeData>

DragDropLabel::DragDropLabel(QWidget *parent):
    QLabel(parent){}

void DragDropLabel::mousePressEvent(QMouseEvent *event)
{
    if (event->buttons() == Qt::LeftButton)
    {
        setStyleSheet("color: blue;");
        QDrag *drag = new QDrag(this);
        drag->setMimeData(new QMimeData());
        drag->exec(Qt::MoveAction);
        QRect rect = ((QWidget*)parent())->rect();
        QPoint pos = ((QWidget*)parent())->mapFromGlobal(QCursor::pos());
        if (!rect.contains(pos))
            setVisible(false);
        setStyleSheet("color: black;");
        QString s = "";
        for (auto child = parent()->children().begin();
             child < parent()->children().begin() + 4; child++)
        {
            if (((DragDropLabel*)(*child))->isVisible())
                return;
            s += ((DragDropLabel*)(*child))->text();
        }
        if (s == "")
            return;
        emit buttonChanged();
    }
}

void DragDropLabel::dragEnterEvent(QDragEnterEvent *event)
{
    event->accept();
    setStyleSheet(styleSheet().append("background-color: yellow;"));
}

void DragDropLabel::dropEvent(QDropEvent *event)
{
    DragDropLabel *src = (DragDropLabel*)event->source();
    setStyleSheet(styleSheet().append("background-color: #f0f0f0;"));
    if (this == src)
        return;
    if (src->tag() > tag())
    {
        src->move(pos());
        setVisible(false);
    }
    else
        src->setVisible(false);
}

void DragDropLabel::dragLeaveEvent(QDragLeaveEvent*)
{
    setStyleSheet(styleSheet().append("background-color: #f0f0f0;"));
}

File: Python/I/11-ZOO/6/dragdroplabel.py
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *

class DragDropLabel(QLabel):
    tag = -1
    
    def mousePressEvent(self, event):
        if event.buttons() == Qt.LeftButton:
            self.setStyleSheet('color: blue;')
            drag = QDrag(self)
            drag.setMimeData(QMimeData())
            drag.exec(Qt.MoveAction)
            rect = self.parent().rect()
            pos = self.parent().mapFromGlobal(QCursor.pos())
            if not rect.contains(pos):
                self.setVisible(False)
            self.setStyleSheet('color: black;')
            s = ''
            for child in self.parent().children()[0:4]:
                if child.isVisible():
                # один из зверей на свободе
                    return
                s += child.text()
            if s == '':
                # все клетки пусты
                return
            self.parent().pushButton.setText('Зоопарк открыт')
            self.parent().pushButton.setStyleSheet('color: green;')
            self.parent().pushButton.setIcon(
                self.style().standardIcon(QStyle.SP_DialogApplyButton))

    def dragEnterEvent(self, event):
        event.accept()
        self.setStyleSheet(
            self.styleSheet() + 'background-color: yellow;')
        
    def dropEvent(self, event):
        src = event.source()
        self.setStyleSheet(
            self.styleSheet() + 'background-color: #F0F0F0;')
        if self == src:
            return
        if src.tag > self.tag:
            src.move(self.pos())
            self.setVisible(False)
        else:
            src.setVisible(False)

    def dragLeaveEvent(self, event):
        self.setStyleSheet(
            self.styleSheet() + 'background-color: #F0F0F0;')
